File: kotlin/advent/src/main/kotlin/org/gristle/pdxpuzzles/advent/y2023/Y23D15.kt
package org.gristle.pdxpuzzles.advent.y2023

import org.gristle.pdxpuzzles.advent.utilities.Day

class Y23D15(input: String) : Day {

    private val initialization = input.split(',')
    
    private fun String.toHash() = fold(0) { acc, c -> (acc + c.code) * 17 % 256 }
    
    override fun part1() = initialization.sumOf { it.toHash() }

    override fun part2(): Int {
        val boxes = List(256) { mutableMapOf<String, Int>()}
        initialization.forEach { step ->
            val (label, operation) = step.partition { it.isLetter() }
            val box = label.toHash()
            if (operation == "-") {
                boxes[box].remove(label)
            } else {
                boxes[box][label] = operation.last().digitToInt() 
            }
        }
        return boxes
            .mapIndexed { boxIndex, box ->
                box.values.foldIndexed(0) { lensIndex, acc, focalLength -> 
                    acc + (boxIndex + 1) * (lensIndex + 1) * focalLength 
                }
            }.sum()
    }
}

fun main() = Day.runDay(Y23D15::class)

//    Class creation: 3ms
//    Part 1: 505427 (1ms)
//    Part 2: 243747 (7ms)
//    Total time: 12ms

@Suppress("unused")
private val sampleInput = listOf(
    """rn=1,cm-,qp=3,cm=2,qp-,pc=4,ot=9,ab=5,pc-,pc=6,ot=7""",
)

File: rust/advent/src/bin/2023/y23d15.rs
use std::collections::HashMap;

use advent::utilities::get_input::get_input;
use itertools::Itertools;
use utilities::structs::stopwatch::{ReportDuration, Stopwatch};

type Input<'a> = &'a str;
type Output = usize;

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let input = get_input(23, 15).unwrap();
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(&input), stopwatch.lap().report());
    println!("2. {} ({})", part2(&input), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

fn hash_of(step: &str) -> usize {
    step.as_bytes()
        .iter()
        .fold(0, |acc, &c| (acc + c as usize) * 17 % 256)
}

fn parse_input(input: Input) -> impl Iterator<Item = &str> {
    input.trim_end().split(',')
}

fn part1(input: Input) -> Output {
    parse_input(input).map(hash_of).sum()
}

fn part2(input: Input) -> Output {
    let mut boxes: [HashMap<&str, (usize, usize)>; 256] = std::array::from_fn(|_| HashMap::new());
    parse_input(input).enumerate().for_each(|(index, step)| {
        let (label, operation) = step.split_once(|c: char| !c.is_ascii_alphabetic()).unwrap();
        let box_number = hash_of(label);
        if operation == "" {
            boxes[box_number].remove(label);
        } else {
            let digit = operation.parse().unwrap();
            let index = if let Some((i, _)) = boxes[box_number].get(label) {
                *i
            } else {
                index
            };
            boxes[box_number].insert(label, (index, digit));
        }
    });
    boxes
        .into_iter()
        .enumerate()
        .map(|(box_index, lens_box)| {
            lens_box
                .values()
                .sorted_by(|(a, _), (b, _)| Ord::cmp(&a, &b))
                .enumerate()
                .fold(0, |acc, (lens_index, (_, focal_length))| {
                    acc + (box_index + 1) * (lens_index + 1) * focal_length
                })
        })
        .sum()
}

#[test]
fn default() {
    let input = get_input(23, 15).unwrap();
    assert_eq!(505427, part1(&input));
    assert_eq!(243747, part2(&input));
}

// Input parsed (21μs)
// 1. 505427 (68μs)
// 2. 243747 (244μs)
// Total: 336μs
